File: core/src/test/kotlin/com/example/toyracers/race/RaceRulesTest.kt
package com.example.toyracers.race

import com.example.toyracers.track.TrackLoader
import com.example.toyracers.track.TrackPoint
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Test

class RaceRulesTest {
    private val track = TrackLoader().load()

    @Test
    fun `checkpoints only count in order`() {
        val rules = RaceRules(track)
        val progress = RaceProgress()

        crossCheckpoint(rules, progress, 1)
        assertEquals(0, progress.currentCheckpointIndex)

        crossCheckpoint(rules, progress, 0)
        assertEquals(1, progress.currentCheckpointIndex)
        crossCheckpoint(rules, progress, 2)
        assertEquals(1, progress.currentCheckpointIndex)
        crossCheckpoint(rules, progress, 1)
        assertEquals(2, progress.currentCheckpointIndex)
    }

    @Test
    fun `backward crossing and repeated gate do not advance progress`() {
        val rules = RaceRules(track)
        val progress = RaceProgress()

        rules.update(
            progress,
            previousPosition = point(35f, 13f),
            currentPosition = point(35f, 11f),
            deltaSeconds = 0f,
        )
        assertEquals(0, progress.currentCheckpointIndex)

        crossCheckpoint(rules, progress, 0)
        crossCheckpoint(rules, progress, 0)
        assertEquals(1, progress.currentCheckpointIndex)
    }

    @Test
    fun `finish line requires every checkpoint and valid direction`() {
        val rules = RaceRules(track)
        val progress = RaceProgress()

        crossFinish(rules, progress)
        assertEquals(0, progress.completedLaps)

        completeCheckpoints(rules, progress)
        crossFinishBackward(rules, progress)
        assertEquals(0, progress.completedLaps)
        crossFinish(rules, progress)

        assertEquals(1, progress.completedLaps)
        assertEquals(0, progress.currentCheckpointIndex)
    }

    @Test
    fun `three laps finish race and record deterministic timing`() {
        val rules = RaceRules(track, requiredLaps = 3)
        val progress = RaceProgress()

        advanceTime(rules, progress, 12f)
        completeLap(rules, progress)
        advanceTime(rules, progress, 10f)
        completeLap(rules, progress)
        advanceTime(rules, progress, 14f)
        completeLap(rules, progress)

        assertTrue(progress.finished)
        assertEquals(3, progress.completedLaps)
        assertEquals(36f, progress.totalRaceTime, TOLERANCE)
        assertEquals(10f, progress.bestLapTime ?: -1f, TOLERANCE)
        assertEquals(1, progress.finishPosition)

        advanceTime(rules, progress, 5f)
        assertEquals(36f, progress.totalRaceTime, TOLERANCE)
    }

    @Test
    fun `respawn movement cannot count a checkpoint`() {
        val rules = RaceRules(track)
        val progress = RaceProgress()

        rules.update(
            progress,
            previousPosition = point(35f, 11f),
            currentPosition = point(35f, 13f),
            deltaSeconds = 1f,
            allowProgress = false,
        )

        assertEquals(0, progress.currentCheckpointIndex)
        assertEquals(1f, progress.totalRaceTime, TOLERANCE)
        assertFalse(progress.finished)
        assertNull(progress.finishPosition)
    }

    @Test(expected = IllegalArgumentException::class)
    fun `negative simulation time is rejected`() {
        RaceRules(track).update(
            RaceProgress(),
            previousPosition = point(0f, 0f),
            currentPosition = point(0f, 0f),
            deltaSeconds = -0.1f,
        )
    }

    private fun completeLap(
        rules: RaceRules,
        progress: RaceProgress,
    ) {
        completeCheckpoints(rules, progress)
        crossFinish(rules, progress)
    }

    private fun completeCheckpoints(
        rules: RaceRules,
        progress: RaceProgress,
    ) {
        track.checkpoints.indices.forEach { crossCheckpoint(rules, progress, it) }
    }

    private fun crossCheckpoint(
        rules: RaceRules,
        progress: RaceProgress,
        index: Int,
    ) {
        when (index) {
            0 -> rules.update(progress, point(35f, 11f), point(35f, 13f), 0f)
            1 -> rules.update(progress, point(22f, 18f), point(19f, 18f), 0f)
            2 -> rules.update(progress, point(7f, 13f), point(7f, 11f), 0f)
        }
    }

    private fun crossFinish(
        rules: RaceRules,
        progress: RaceProgress,
    ) {
        rules.update(progress, point(19f, 6f), point(22f, 6f), 0f)
    }

    private fun crossFinishBackward(
        rules: RaceRules,
        progress: RaceProgress,
    ) {
        rules.update(progress, point(22f, 6f), point(19f, 6f), 0f)
    }

    private fun advanceTime(
        rules: RaceRules,
        progress: RaceProgress,
        seconds: Float,
    ) {
        rules.update(progress, point(6f, 6f), point(6f, 6f), seconds)
    }

    private fun point(
        x: Float,
        y: Float,
    ) = TrackPoint(x, y)

    private companion object {
        const val TOLERANCE = 0.001f
    }
}

File: dart/test/race_rules_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:toy_racers/simulation.dart';

void main() {
  late Track track;
  late RaceRules rules;

  setUp(() {
    track = _track();
    rules = RaceRules(track, requiredLaps: 2);
  });

  test('advances only the next checkpoint and rejects invalid crossings', () {
    final progress = RaceProgress();

    _crossCheckpoint(rules, progress, 1);
    _crossCheckpointBackward(rules, progress, 0);
    expect(progress.currentCheckpointIndex, 0);

    _crossCheckpoint(rules, progress, 0);
    _crossCheckpoint(rules, progress, 0);
    expect(progress.currentCheckpointIndex, 1);

    _crossCheckpoint(rules, progress, 1);
    expect(progress.currentCheckpointIndex, 2);
  });

  test(
    'records lap, race, and best-lap timers before deterministic finish',
    () {
      final progress = RaceProgress();

      _advanceTime(rules, progress, 12);
      _completeLap(rules, progress);
      _advanceTime(rules, progress, 10);
      _completeLap(rules, progress);

      expect(progress.completedLaps, 2);
      expect(progress.totalRaceTime, 22);
      expect(progress.bestLapTime, 10);
      expect(progress.finished, isTrue);
      expect(progress.finishPosition, 1);

      _advanceTime(rules, progress, 5);
      expect(progress.totalRaceTime, 22);
    },
  );

  test(
    'suppressed respawn movement advances time but not checkpoint progress',
    () {
      final progress = RaceProgress();

      rules.update(
        progress: progress,
        previousPosition: TrackPoint(19, 50),
        currentPosition: TrackPoint(21, 50),
        deltaSeconds: 1,
        allowProgress: false,
      );

      expect(progress.currentCheckpointIndex, 0);
      expect(progress.totalRaceTime, 1);
      expect(progress.finished, isFalse);
      expect(progress.finishPosition, isNull);
    },
  );

  test('continues restored finish ordering and rejects invalid ordering', () {
    final restored = RaceProgress(finished: true, finishPosition: 2);
    final next = RaceProgress();

    rules.synchronizeFinishOrdering(<RaceProgress>[restored]);
    _completeLap(rules, next);
    _completeLap(rules, next);

    expect(next.finishPosition, 3);
    expect(
      () => rules.synchronizeFinishOrdering(<RaceProgress>[
        RaceProgress(finished: true),
      ]),
      throwsArgumentError,
    );
    expect(
      () => rules.synchronizeFinishOrdering(<RaceProgress>[
        RaceProgress(finished: true, finishPosition: 4),
        RaceProgress(finished: true, finishPosition: 4),
      ]),
      throwsArgumentError,
    );
  });

  test('resets finish ordering for a fresh race', () {
    final firstRaceProgress = RaceProgress();
    _completeLap(rules, firstRaceProgress);
    _completeLap(rules, firstRaceProgress);
    expect(firstRaceProgress.finishPosition, 1);

    rules.resetFinishOrdering();

    final nextRaceProgress = RaceProgress();
    _completeLap(rules, nextRaceProgress);
    _completeLap(rules, nextRaceProgress);
    expect(nextRaceProgress.finishPosition, 1);
  });

  test('requires positive laps and non-negative simulation time', () {
    expect(() => RaceRules(track, requiredLaps: 0), throwsArgumentError);
    expect(
      () => rules.update(
        progress: RaceProgress(),
        previousPosition: TrackPoint(0, 0),
        currentPosition: TrackPoint(0, 0),
        deltaSeconds: -0.1,
      ),
      throwsArgumentError,
    );
  });
}

void _advanceTime(RaceRules rules, RaceProgress progress, double seconds) {
  rules.update(
    progress: progress,
    previousPosition: TrackPoint(10, 10),
    currentPosition: TrackPoint(10, 10),
    deltaSeconds: seconds,
  );
}

void _completeLap(RaceRules rules, RaceProgress progress) {
  _crossCheckpoint(rules, progress, 0);
  _crossCheckpoint(rules, progress, 1);
  rules.update(
    progress: progress,
    previousPosition: TrackPoint(50, 50),
    currentPosition: TrackPoint(52, 50),
    deltaSeconds: 0,
  );
}

void _crossCheckpoint(RaceRules rules, RaceProgress progress, int index) {
  final y = index == 0 ? 50.0 : 70.0;
  rules.update(
    progress: progress,
    previousPosition: TrackPoint(19, y),
    currentPosition: TrackPoint(21, y),
    deltaSeconds: 0,
  );
}

void _crossCheckpointBackward(
  RaceRules rules,
  RaceProgress progress,
  int index,
) {
  final y = index == 0 ? 50.0 : 70.0;
  rules.update(
    progress: progress,
    previousPosition: TrackPoint(21, y),
    currentPosition: TrackPoint(19, y),
    deltaSeconds: 0,
  );
}

Track _track() {
  final bounds = TrackRectangle(0, 0, 100, 100);
  return Track.fromDefinition(
    id: 'race-rules-test',
    name: 'RACE RULES TEST',
    worldBounds: bounds,
    cameraBounds: bounds,
    outerBoundary: bounds,
    backgroundSurface: SurfaceType.asphalt,
    startLine: StartLine(
      bounds: TrackRectangle(50, 40, 2, 20),
      forwardX: 1,
      forwardY: 0,
    ),
    checkpoints: <Checkpoint>[
      Checkpoint(
        order: 0,
        gate: TrackSegment(TrackPoint(20, 40), TrackPoint(20, 60)),
        forwardX: 1,
        forwardY: 0,
      ),
      Checkpoint(
        order: 1,
        gate: TrackSegment(TrackPoint(20, 60), TrackPoint(20, 80)),
        forwardX: 1,
        forwardY: 0,
      ),
    ],
    startGrid: <StartGridPosition>[
      StartGridPosition(position: TrackPoint(10, 50), rotationDegrees: 0),
    ],
    racingLine: <TrackPoint>[
      TrackPoint(10, 50),
      TrackPoint(20, 50),
      TrackPoint(50, 50),
    ],
  );
}
